compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -527,6 +534,160 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+    /// - It wasn't completely captured by the closure, **and**
+    /// - One of the paths captured does not implement all the auto-traits its root variable
+    ///   implements.
+    fn compute_2229_migrations_for_trait(
+        &self,
+        min_captures: Option<&ty::RootVariableMinCaptureList<'tcx>>,
+        closure_clause: hir::CaptureBy,
+        var_hir_id: hir::HirId,
+    ) -> Option<(hir::HirId, FxHashSet<&str>)> {
+        use crate::rustc_middle::ty::ToPredicate;
+        use crate::rustc_middle::ty::WithConstness;
+        use rustc_trait_selection::traits::query::evaluate_obligation::InferCtxtExt;
+
+        let ty = self.infcx.resolve_vars_if_possible(self.node_ty(var_hir_id));
+
+        let tcx = self.infcx.tcx;
+
+        let cause = ObligationCause::misc(self.tcx.hir().span(var_hir_id), self.body_id);
+
+        // <ty as Clone>
+        let clone_trait_ref = TraitRef {
+            def_id: tcx.lang_items().clone_trait().unwrap(),
+            substs: tcx.mk_substs_trait(ty, &[]),
+        };
+
+        let clone_obligation = Obligation::new(
+            cause.clone(),
+            self.param_env,
+            clone_trait_ref.without_const().to_predicate(tcx),
+        );
+
+        // <ty as Sync>
+        let sync_trait_ref = TraitRef {
+            def_id: tcx.lang_items().sync_trait().unwrap(),
+            substs: tcx.mk_substs_trait(ty, &[]),
+        };
+
+        let sync_obligation = Obligation::new(
+            cause.clone(),
+            self.param_env,
+            sync_trait_ref.without_const().to_predicate(tcx),
+        );
+
+        // <ty as Unpin>
+        let unpin_trait_ref = TraitRef {
+            def_id: tcx.lang_items().unpin_trait().unwrap(),
+            substs: tcx.mk_substs_trait(ty, &[]),
+        };
+
+        let unpin_obligation = Obligation::new(
+            cause.clone(),
+            self.param_env,
+            unpin_trait_ref.without_const().to_predicate(tcx),
+        );
+
+        if self.infcx.predicate_may_hold(&clone_obligation)
+            && self.infcx.predicate_may_hold(&sync_obligation)
+            && self.infcx.predicate_may_hold(&unpin_obligation)
+        {
+            return None
+        }
+
+        let mut temp_reasons = FxHashSet::default();
+
+        if !self.infcx.predicate_may_hold(&clone_obligation) {
+            temp_reasons.insert("Clone");
+        }
+
+        if !self.infcx.predicate_may_hold(&sync_obligation) {
+            temp_reasons.insert("Sync");
+        }
+
+        if !self.infcx.predicate_may_hold(&unpin_obligation) {
+            temp_reasons.insert("Unpin");
+        }
+
+        let root_var_min_capture_list = if let Some(root_var_min_capture_list) =
+            min_captures.and_then(|m| m.get(&var_hir_id))
+        {
+            root_var_min_capture_list
+        } else {
+            // The upvar is mentioned within the closure but no path starting from it is
+            // used.
+
+            match closure_clause {
```

Comment:
this doesn't matter because this is for move closures, just ignore it

Reply:
Done